File: backend/services/auth_service.go
package services
import (
	"errors"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"cronos-backend/utils"
	"github.com/google/uuid"
)
type AuthService struct {
	UserRepo *repositories.UserRepository
}
func NewAuthService(userRepo *repositories.UserRepository) *AuthService {
	return &AuthService{UserRepo: userRepo}
}
func (s *AuthService) Register(req models.RegisterRequest) (*models.UserResponse, string, error) {
	existing, _ := s.UserRepo.FindByEmail(req.Email)
	if existing != nil && existing.ID != uuid.Nil {
		return nil, "", errors.New("email already registered")
	}
	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return nil, "", errors.New("failed to hash password")
	}
	user := &models.User{
		ID:         uuid.New(),
		Name:       req.Name,
		Email:      req.Email,
		Password:   hashedPassword,
		Phone:      req.Phone,
		Role:       req.Role,
		IsVerified: req.Role == "konsumen", 
	}
	if err := s.UserRepo.Create(user); err != nil {
		return nil, "", errors.New("failed to create user")
	}
	token, err := utils.GenerateToken(user.ID, user.Email, user.Role)
	if err != nil {
		return nil, "", errors.New("failed to generate token")
	}
	resp := user.ToResponse()
	return &resp, token, nil
}
func (s *AuthService) Login(req models.LoginRequest) (*models.UserResponse, string, error) {
	user, err := s.UserRepo.FindByEmail(req.Email)
	if err != nil {
		return nil, "", errors.New("invalid email or password")
	}
	if !utils.CheckPassword(req.Password, user.Password) {
		return nil, "", errors.New("invalid email or password")
	}
	if !user.IsVerified && (user.Role == "petambak" || user.Role == "logistik") {
		return nil, "", errors.New("your account is pending verification by admin")
	}
	token, err := utils.GenerateToken(user.ID, user.Email, user.Role)
	if err != nil {
		return nil, "", errors.New("failed to generate token")
	}
	resp := user.ToResponse()
	return &resp, token, nil
}
func (s *AuthService) GetProfile(userID uuid.UUID) (*models.UserResponse, error) {
	user, err := s.UserRepo.FindByID(userID)
	if err != nil {
		return nil, errors.New("user not found")
	}
	resp := user.ToResponse()
	return &resp, nil
}
func (s *AuthService) UpdateProfile(userID uuid.UUID, req models.UpdateProfileRequest) (*models.UserResponse, error) {
	user, err := s.UserRepo.FindByID(userID)
	if err != nil {
		return nil, errors.New("user not found")
	}
	if req.Name != "" {
		user.Name = req.Name
	}
	if req.Phone != "" {
		user.Phone = req.Phone
	}
	if req.Address != "" {
		user.Address = req.Address
	}
	if req.Avatar != "" {
		user.Avatar = req.Avatar
	}
	if err := s.UserRepo.Update(user); err != nil {
		return nil, errors.New("failed to update profile")
	}
	resp := user.ToResponse()
	return &resp, nil
}

File: backend-spring/src/main/java/com/cronos/service/AuthService.java
package com.cronos.service;

import com.cronos.dto.request.LoginRequest;
import com.cronos.dto.request.RegisterRequest;
import com.cronos.dto.request.UpdateProfileRequest;
import com.cronos.dto.response.UserResponse;
import com.cronos.entity.User;
import com.cronos.repository.UserRepository;
import com.cronos.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.Map;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepo;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    public Map<String, Object> register(RegisterRequest req) {
        if (userRepo.findByEmail(req.getEmail()).isPresent()) {
            throw new RuntimeException("email already registered");
        }
        User user = User.builder()
                .name(req.getName())
                .email(req.getEmail())
                .password(passwordEncoder.encode(req.getPassword()))
                .phone(req.getPhone())
                .role(req.getRole())
                .isVerified("konsumen".equals(req.getRole()))
                .balance(BigDecimal.ZERO)
                .build();
        userRepo.save(user);
        String token = jwtTokenProvider.generateToken(user.getId(), user.getEmail(), user.getRole());
        return Map.of("user", UserResponse.fromEntity(user), "token", token);
    }

    public Map<String, Object> login(LoginRequest req) {
        User user = userRepo.findByEmail(req.getEmail())
                .orElseThrow(() -> new RuntimeException("invalid email or password"));
        if (!passwordEncoder.matches(req.getPassword(), user.getPassword())) {
            throw new RuntimeException("invalid email or password");
        }
        if (!user.getIsVerified() && ("petambak".equals(user.getRole()) || "logistik".equals(user.getRole()))) {
            throw new RuntimeException("your account is pending verification by admin");
        }
        String token = jwtTokenProvider.generateToken(user.getId(), user.getEmail(), user.getRole());
        return Map.of("user", UserResponse.fromEntity(user), "token", token);
    }

    public UserResponse getProfile(UUID userId) {
        User user = userRepo.findById(userId).orElseThrow(() -> new RuntimeException("user not found"));
        return UserResponse.fromEntity(user);
    }

    public UserResponse updateProfile(UUID userId, UpdateProfileRequest req) {
        User user = userRepo.findById(userId).orElseThrow(() -> new RuntimeException("user not found"));
        if (req.getName() != null && !req.getName().isEmpty()) user.setName(req.getName());
        if (req.getPhone() != null && !req.getPhone().isEmpty()) user.setPhone(req.getPhone());
        if (req.getAddress() != null && !req.getAddress().isEmpty()) user.setAddress(req.getAddress());
        if (req.getAvatar() != null && !req.getAvatar().isEmpty()) user.setAvatar(req.getAvatar());
        userRepo.save(user);
        return UserResponse.fromEntity(user);
    }
}
